File: Upcoder/Lập trình cơ bản/5.Mảng/Xoắn ốc cơ bản/XoanOcCoBan.py
def main():
    n, m = map(int, input().split())

    a = [[0] * m for _ in range(n)]
    index = 1
    for i in range(n):
        for j in range(m):
            a[i][j] = index
            index += 1

    top, bottom, left, right = 0, n - 1, 0, m - 1
    result = [[0] * m for _ in range(n)]
    count = 0

    while top <= bottom and left <= right:
        # Duyệt từ trái sang phải trên hàng top
        for i in range(left, right + 1):
            result[top][i] = a[count // m][count % m]
            count += 1
        top += 1

        # Duyệt từ trên xuống dưới ở cột right
        for i in range(top, bottom + 1):
            result[i][right] = a[count // m][count % m]
            count += 1
        right -= 1

        # Duyệt từ phải sang trái trên hàng bottom
        if top <= bottom:
            for i in range(right, left - 1, -1):
                result[bottom][i] = a[count // m][count % m]
                count += 1
            bottom -= 1

        # Duyệt từ dưới lên trên ở cột left
        if left <= right:
            for i in range(bottom, top - 1, -1):
                result[i][left] = a[count // m][count % m]
                count += 1
            left += 1

    # In ma trận xoắn ốc
    for row in result:
        print(' '.join(map(str, row)))

if __name__ == "__main__":
    main()

File: Upcoder/Lập trình cơ bản/5.Mảng/Xoắn ốc cơ bản/XoanOcCoBan.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n, m;
    cin >> n >> m;

    int a[n][m];
    int index = 1;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            a[i][j] = index;
            index++;
        }
    }

    int top = 0, bottom = n - 1, left = 0, right = m - 1;
    int result[n][m];
    int count = 0;

    while (top <= bottom && left <= right) {
        // Duyệt từ trái sang phải trên hàng top
        for (int i = left; i <= right; i++) {
            result[top][i] = a[count / m][count % m];
            count++;
        }
        top++;

        // Duyệt từ trên xuống dưới ở cột right
        for (int i = top; i <= bottom; i++) {
            result[i][right] = a[count / m][count % m];
            count++;
        }
        right--;

        // Duyệt từ phải sang trái trên hàng bottom
        if (top <= bottom) {
            for (int i = right; i >= left; i--) {
                result[bottom][i] = a[count / m][count % m];
                count++;
            }
            bottom--;
        }

        // Duyệt từ dưới lên trên ở cột left
        if (left <= right) {
            for (int i = bottom; i >= top; i--) {
                result[i][left] = a[count / m][count % m];
                count++;
            }
            left++;
        }
    }

    // In ma trận xoắn ốc
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cout << result[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}
